providers/dynamodb/shedlock-provider-dynamodb/src/main/java/net/javacrumbs/shedlock/provider/dynamodb/DynamoDBLockProvider.java

```diff
@@ -0,0 +1,238 @@
+ *    "lockedAt" : ISODate("2017-01-07T16:52:03.932Z"),
+ *    "lockedBy" : "host name"
+ * }
+ * </pre>
+ *
+ * <code>lockedAt</code> and <code>lockedBy</code> are just for troubleshooting
+ * and are not read by the code.
+ *
+ * <ol>
+ * <li>
+ * Attempts to insert a new lock record.
+ * </li>
+ * <li>
+ * We will try to update lock record using <code>filter _id == :name AND lock_until &lt;= :now</code>.
+ * </li>
+ * <li>
+ * If the update succeeded, we have the lock. If the update failed (condition check exception)
+ * somebody else holds the lock.
+ * </li>
+ * <li>
+ * When unlocking, <code>lock_until</code> is set to <i>now</i> or <i>lockAtLeastUntil</i> whichever is later.
+ * </li>
+ * </ol>
+ */
+public class DynamoDBLockProvider implements LockProvider {
+    static final String LOCK_UNTIL = "lockUntil";
+    static final String LOCKED_AT = "lockedAt";
+    static final String LOCKED_BY = "lockedBy";
+    static final String ID = "_id";
+
+    static final String DEFAULT_TABLE_NAME = "shedLock";
+
+    private static final String OBTAIN_LOCK_QUERY =
+            "set " + LOCK_UNTIL + " = :lockUntil, " + LOCKED_AT + " = :lockedAt, " + LOCKED_BY + " = :lockedBy";
+    private static final String OBTAIN_LOCK_CONDITION =
+            LOCK_UNTIL + " <= :lockedAt or attribute_not_exists(" + LOCK_UNTIL + ")";
+    private static final String RELEASE_LOCK_QUERY =
+            "set " + LOCK_UNTIL + " = :lockUntil";
+
+    /**
+     * A {@link DateTimeFormatter} like {@link DateTimeFormatter#ISO_INSTANT} with
+     * the exception that it always appends exactly three fractional digits (nano seconds).
+     * <p>
+     * This is required in order to guarantee natural sorting, which enables us to use
+     * <code>&lt;=</code> comparision in queries.
+     *
+     * <pre>
+     * 2018-12-07T12:30:37.000Z
+     * 2018-12-07T12:30:37.810Z
+     * 2018-12-07T12:30:37.819Z
+     * 2018-12-07T12:30:37.820Z
+     * </pre>
+     *
+     * When using variable fractional digit count as done in {@link DateTimeFormatter#ISO_INSTANT ISO_INSTANT}
+     * and {@link DateTimeFormatter#ISO_OFFSET_DATE_TIME ISO_OFFSET_DATE_TIME} the following sorting
+     * occurs:
+     *
+     * <pre>
+     * 2018-12-07T12:30:37.819Z
+     * 2018-12-07T12:30:37.81Z
+     * 2018-12-07T12:30:37.820Z
+     * 2018-12-07T12:30:37Z
+     * </pre>
+     *
+     * @see <a href="https://stackoverflow.com/a/5098252">natural sorting of ISO 8601 time format</a>
+     */
+    private static final DateTimeFormatter formatter = new DateTimeFormatterBuilder()
+            .parseCaseInsensitive()
+            .appendInstant(3)
+            .toFormatter();
+
+    private final String hostname;
+    private final Table table;
+
+    /**
+     * Uses DynamoDB to coordinate locks stored in table {@value DEFAULT_TABLE_NAME}.
+     *
+     * @param dynamodb  DynamoDB to be used
+     */
+    public DynamoDBLockProvider(AmazonDynamoDB dynamodb) {
```

Comment:
I am wondering if the constructor with default table name is useful. In Mongo it is, the collection is created automatically. In JDBC less so, but the excuse is that you create a table by provided DDL statement so it may be useful. But in Dynamo I assume you will create the table using Java code where you can specify some constant so maybe it does not make much sense to have it here.

Reply:
What about adding a constructor with Table as a single parameter?

Reply:
Fair points. Do you mean DynamoDB‘s Table instance?

Reply:
Yes, please